### PR TITLE
[New Feature][Solace Integration-01] Add environment configs and properties to support external gateways

### PR DESCRIPTION
**Pull request by Chamindu36**
### Purpose
Add new parameters to environment models to support third-party gateway environments. Add the ability to get the third-party gateway environment configurations from _deployment.toml_ file. 

### Goal
Fixes : 
https://github.com/wso2/product-apim/issues/11649
https://github.com/wso2/product-apim/issues/11725

### Unit Tests

- testEnvironmentsConfigProvided - Test whether the provided gateway config is read and extracted correctly.
- testEnvironmentsConfigWithAdditionalProperties - Test whether the provided gateway config with additional properties is read and extracted correctly.

### CODE REVIEW

File: components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/Environment.java [hunk 1]
```diff
@@ -49,6 +51,8 @@
     private String description;
     private boolean isReadOnly;
     private List<VHost> vhosts = new ArrayList<>();
+    private String provider;
+    private Map<String, String> additionalProperties = new HashMap<>();
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.api/src/main/java/org/wso2/carbon/apimgt/api/model/Environment.java:55]: What type of additional properties ? Is this related solace ?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 2]
```diff
@@ -626,6 +523,135 @@ private void readChildElements(OMElement serverConfig,
         }
     }
 
+    /**
+     * Set property values for each gateway environments defined in the api-manager.xml config file
+     *
+     * @param environmentElem OMElement of a single environment in the gateway environments list
+     */
+    void setEnvironmentConfig(OMElement environmentElem) throws APIManagementException {
+
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java:532]: remove new line.check in other places as well

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 3]
```diff
@@ -626,6 +523,135 @@ private void readChildElements(OMElement serverConfig,
...
+        environment.setType(environmentElem.getAttributeValue(new QName("type")));
+        String showInConsole = environmentElem.getAttributeValue(new QName("api-console"));
+        if (showInConsole != null) {
+            environment.setShowInConsole(Boolean.parseBoolean(showInConsole));
+        } else {
+            environment.setShowInConsole(true);
+        }
+        String isDefault = environmentElem.getAttributeValue(new QName("isDefault"));
+        if (isDefault != null) {
+            environment.setDefault(Boolean.parseBoolean(isDefault));
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java:543]: Can't we add more info to Default variable like set default what ? is it make sense ?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 4]
```diff
@@ -626,6 +523,135 @@ private void readChildElements(OMElement serverConfig,
...
+        } else {
+            environment.setDefault(false);
+        }
+        environment.setName(APIUtil.replaceSystemProperty(
+                environmentElem.getFirstChildWithName(new QName(
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java:548]: fix formatting we have more space to include in one line. Fix all the places

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 5]
```diff
@@ -626,6 +523,135 @@ private void readChildElements(OMElement serverConfig,
...
+        }
+        environment.setServerURL(APIUtil.replaceSystemProperty(
+                environmentElem.getFirstChildWithName(new QName(
+                        APIConstants.API_GATEWAY_SERVER_URL)).getText()));
+        environment.setUserName(APIUtil.replaceSystemProperty(
+                environmentElem.getFirstChildWithName(new QName(
+                        APIConstants.API_GATEWAY_USERNAME)).getText()));
+        OMElement passwordElement = environmentElem.getFirstChildWithName(new QName(
+                APIConstants.API_GATEWAY_PASSWORD));
+        String value = MiscellaneousUtil.resolve(passwordElement, secretResolver);
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java:564]: Have a better name than value

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 6]
```diff
@@ -626,6 +523,135 @@ private void readChildElements(OMElement serverConfig,
...
+            String wsEp = APIUtil.replaceSystemProperty(vhostElem.getFirstChildWithName(new QName(
+                    APIConstants.API_GATEWAY_VIRTUAL_HOST_WS_ENDPOINT)).getText());
+            String wssEp = APIUtil.replaceSystemProperty(vhostElem.getFirstChildWithName(new QName(
+                    APIConstants.API_GATEWAY_VIRTUAL_HOST_WSS_ENDPOINT)).getText());
+            String webSubHttpEp = APIUtil.replaceSystemProperty(vhostElem.getFirstChildWithName(new QName(
+                    APIConstants.API_GATEWAY_VIRTUAL_HOST_WEBSUB_HTTP_ENDPOINT)).getText());
+            String webSubHttpsEp = APIUtil.replaceSystemProperty(vhostElem.getFirstChildWithName(new QName(
+                    APIConstants.API_GATEWAY_VIRTUAL_HOST_WEBSUB_HTTPS_ENDPOINT)).getText());
+
+                        /*
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java:617]: fix formatting

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 7]
```diff
@@ -626,6 +523,135 @@ private void readChildElements(OMElement serverConfig,
...
+                String propValue = propertyElem.getText();
+                additionalProperties.put(propName, propValue);
+            }
+        }
+        environment.setAdditionalProperties(additionalProperties);
+
+        if (!apiGatewayEnvironments.containsKey(environment.getName())) {
+            apiGatewayEnvironments.put(environment.getName(), environment);
+        } else {
+                        /*
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java:646]: Fix formatting check all places

Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10839/commits/20106d3b631c172110c54b848257432bf17bb19d

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java (hunk 5 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10839/commits/20106d3b631c172110c54b848257432bf17bb19d

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java (hunk 2 above)
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java]: Fixed with https://github.com/wso2/carbon-apimgt/pull/10839/commits/20106d3b631c172110c54b848257432bf17bb19d

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java [hunk 8]
```diff
@@ -626,6 +523,120 @@ private void readChildElements(OMElement serverConfig,
...
+        OMElement properties = environmentElem.getFirstChildWithName(new
+                QName(APIConstants.API_GATEWAY_ADDITIONAL_PROPERTIES));
+        Map<String, String> additionalProperties = new HashMap<>();
+        if (properties != null) {
+            Iterator gatewayAdditionalProperties = properties.getChildrenWithLocalName
+                    (APIConstants.API_GATEWAY_ADDITIONAL_PROPERTY);
+            while (gatewayAdditionalProperties.hasNext()) {
+                OMElement propertyElem = (OMElement) gatewayAdditionalProperties.next();
+                String propName = propertyElem.getAttributeValue(new QName("name"));
+                String propValue = propertyElem.getText();
```
Comment (tharindu1st) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java:623]: check the value from vault same as https://github.com/wso2/carbon-apimgt/pull/10839/files#diff-6ad5a30e0da551268318842e70c7a6e235b4e71a5985b57b662723142c8d9ff9R556-R557

Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/main/java/org/wso2/carbon/apimgt/impl/APIManagerConfiguration.java]: Fixed with 8bcaf927123b460db95dcb2223becbdbff7946eb

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIManagerConfigurationTest.java [hunk 9]
```diff
@@ -53,4 +55,76 @@ public void testPublisherPropertiesUndefined() throws XMLStreamException {
         Assert.assertTrue(extractedProperties.isEmpty());
 
     }
+
+    @Test
+    public void testEnvironmentsConfigProvided() throws XMLStreamException, APIManagementException {
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIManagerConfigurationTest.java:60]: is there any mandatory parameters for this config ? Did we write tests to validate those ?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIManagerConfigurationTest.java [hunk 10]
```diff
@@ -53,4 +55,76 @@ public void testPublisherPropertiesUndefined() throws XMLStreamException {
...
+        String envConfig = "<Environment type=\"hybrid\" api-console=\"true\" isDefault=\"true\">\n" +
+                "                <Name>Default</Name>\n" +
+                "                <DisplayName></DisplayName>\n" +
+                "                <Description>This is a hybrid gateway that handles both production and sandbox token traffic.</Description>\n" +
+                "                <!-- Server URL of the API gateway -->\n" +
+                "                <ServerURL>https://localhost:9440/services/</ServerURL>\n" +
```
Comment (dnwick) [inline components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIManagerConfigurationTest.java:66]: what happens if this is not a string in url format ?

File: components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIManagerConfigurationTest.java [hunk 11]
```diff
@@ -53,4 +55,76 @@ public void testPublisherPropertiesUndefined() throws XMLStreamException {
...
+                "                <GatewayEndpoint>https://localhost:9440,http://localhost:9440</GatewayEndpoint>\n" +
+                "                <!-- Additional properties for External Gateways -->\n" +
+                "                <!-- Endpoint URLs of the WebSocket APIs hosted in this API Gateway -->\n" +
+                "                <GatewayWSEndpoint>ws://localhost:9099,wss://localhost:8099</GatewayWSEndpoint>\n" +
+                "                <!-- Endpoint URLs of the WebSub APIs hosted in this API Gateway -->\n" +
+                "                <GatewayWebSubEndpoint>http://localhost:9021,https://localhost:8021</GatewayWebSubEndpoint>\n" +
+                "                <Properties>\n" +
+                "                    <Property name=\"Organization\">WSO2</Property>\n" +
+                "                    <Property name=\"DisplayName\">Development Environment</Property>\n" +
+                "                    <Property name=\"DevAccountName\">dev-1</Property>\n" +
```
[the remark this answers is not in the document]
Comment (Chamindu36) [reply components/apimgt/org.wso2.carbon.apimgt.impl/src/test/java/org/wso2/carbon/apimgt/impl/APIManagerConfigurationTest.java:117]: Since these are additional properties defined as key-value pairs, we blindly add the property name and value. Hence, we cannot do any validations there to check user errors.
